### ISSUE TITLE
Atomic Bomberman is lagging

### ISSUE THREAD

**Issue by Ceotom** [solemnwarning/ipxwrapper#14]
We have 2 players with following setup:
1 player: Hosts DOSBox ipxnet server (using DOSBox-X)
2 player: Connects to 1 player's ipxnet server 

Both running Win10, who hosts in-game doesn't matter. Players teleports continuously and doesn't properly sync some events, like bomb placement, power-up pickups. Bombs can appear at wrong places or too late, power-ups may re-appear at same place after being picked up and etc.
Video from 1 player side (red in-game):
https://github.com/solemnwarning/ipxwrapper/assets/36301121/816715b9-57d1-4ac1-a1c6-39fa43489b80
You can notice how 2 player walks jerky
17s: 1 player pickups disease that forces player to set bombs continuously as soon as possible, but at 19s it reappears below bomb
25s: For some reason 2 bomb unexpectedly appears behind 1 player which is impossible normally (he can set only 1 bomb)

Logs with debugging messages (we didn't wait for the game to start because it takes VERY LONG to start with debugging enabled and we quit on "Receiving level data"):
[1player.log](https://github.com/solemnwarning/ipxwrapper/files/13402471/1player.log)
[2player.log](https://github.com/solemnwarning/ipxwrapper/files/13402472/2player.log)

During test with VMware VMs with this setup:

Host machine: Hosts DOSBox ipxnet server on Win10
VM1 and 2: Connects to host's ipxnet server using host-only adapter, running WinXP due to performance reasons, 2 VMs with Win10 are too slow.

I'm found that those issues appears only if simulate packet loss, ping just add delays until you action appear on other side.

Pinging showed that we has ~38ms ping and 0% packet loss. Looks like ipxwrapper drops packets?

**Comment by solemnwarning** [comment]
Hello,

As you can probably tell from looking at the issues list on here, I've been struggling with some people encountering lag or general weirdness but never managed to pin down the cause.

In your testing with VMWare, are both VMs running on the same host, on the same network or across the Internet?

> I'm found that those issues appears only if simulate packet loss, ping just add delays until you action appear on other side.

What are you using to simulate packet loss?

Thanks

**Comment by solemnwarning** [comment]
So, the logs you attached were from an actual play session, not the VMWare packet loss test, right? They're suggesting packet loss may be going on (1player transmitted 3,182 packets vs. 2player receiving 1,242 packets), but some of that could be down to them not quite lining up time wise (i.e. one game being exited before the other).

**Comment by solemnwarning** [comment]
Player 1 sent 3,182 packets and Player 2 received 1,242
Player 2 sent 1,007 and Player 1 received 976

**Comment by solemnwarning** [comment]
I spent a bit of time fiddling earlier. Bomberman is sending lots of small broadcast packets which seem to come through okay with the native IPX driver on XP, swapping in IPXWrapper degrades performance slightly, but no obvious cause. Switching to DOSBox encapsulation degrades it badly.

Sending lots of UDP packets like this isn't a "done" thing, and as a result the IP stack of any involved machine could be dropping, throttling or awkwardly shaping the traffic.

Depending on how bomberman processes traffic, I might be able to introduce a feature to coalesce frequent-but-small packets into fewer batched ones, which *might* help.

First though, I'm finishing off a profiling feature I was working off before which will hopefully clear things up.

**Comment by solemnwarning** [comment]
The issue isn't the packets themselves being small, but the sheer number of them. Bomberman is sending 400+ (broadcast) packets every second when in a network game (most of them <50 bytes), which is resulting in it spending a lot of time sending/receiving packets. I don't have profiling data to back this up, but I can only assume the original IPX driver in Windows was a lot more simplistic than the IP stack in modern Windows.

It seems to be spending around 1s inside the IPXWrapper `sendto()` function out of every 10 seconds, a good chunk of which is inside the OS `sendto()` call.

I experimented a bit with using `WSASendTo()` to send the packets using overlapped I/O instead, which theoretically offers better performance since the thread can return and be ready to queue up another packet before the first one has been fully committed into the socket buffer, but that actually seems to make things even slower, probably because of how stupidly tiny and numerous these packets are.

I'll have to try and find more optimisations that can be made elsewhere.

**Comment by solemnwarning** [comment]
So there are still places within IPXWrapper where I _might_ be able to squeeze a bit more performance out where certain APIs are being used in terrible ways by Atomic Bomberman, but I still think the root cause here is the sheer number of packets.

I've attached a prototype build of IPXWrapper which batches and merges packets when the application starts sending them at ridiculous rates (~90% packet count reduction during AB gameplay). Can you and your friend give it a try and report how it works?

There is now a profiling option in ipxconfig you can turn on to gather useful statistics for this which shouldn't slow the game down like the full debug log does.

[ipxwrapper-git.zip](https://github.com/solemnwarning/ipxwrapper/files/13468084/ipxwrapper-git.zip)

**Comment by solemnwarning** [comment]
Well that raises more questions than it answers.

Going by the Prof logs there, the packet coalescing I implemented isn't actually kicking in (there should be a message along the lines of "High send rate to xxx detected, coalescing future packets"), so nothing else in my build should be significantly faster.

The stats in the log back this up too:

```
[988646.95, thread 5088] IPX sockets sent 31939 packets (1089730 bytes)
[988646.98, thread 5088] IPX sockets received 3009 packets (104774 bytes)
[988647.03, thread 5088] UDP sockets sent 31939 packets (2047900 bytes)
[988647.06, thread 5088] UDP sockets received 3009 packets (195044 bytes)
```

If the packet coalescing was working, the UDP socket stats would show far fewer packets than the IPX socket stats. The counters suggest there is a lot of packet loss going on, which isn't too surprising at the packet rates above.

**Comment by solemnwarning** [comment]
I haven't been able to come up with any reasons why the packet coalescing might not kick in, so could you try this build out and send me the logs with profiling enabled? Hopefully something in there will point at why it isn't working.

Thanks

[ipxwrapper-git.zip](https://github.com/solemnwarning/ipxwrapper/files/13477786/ipxwrapper-git.zip)

**Comment by solemnwarning** [comment]
Unfortunately, the Atomic Bomberman network code just isn't very good. The best I can do is tweak IPXWrapper to reduce the chances of packets getting lost and improve the speed of the calls it is (over)using - as you've seen, when packets are lost it can go out of sync and makes no attempt at correcting the state between players. The walking animation not playing isn't even specific to IPXWrapper - that happens when using the native Windows IPX driver on XP too.

I've made a few more optimisations in this build which _might_ help a little bit, let me know how you get on.

[ipxwrapper-git.zip](https://github.com/solemnwarning/ipxwrapper/files/13538886/ipxwrapper-git.zip)

**Comment by insellium** [comment]
Hi. I was testing Atomic Bomberman with Ceotom a month or two ago.
How is the fixing process going?

**Comment by solemnwarning** [comment]
@insellium I've not got any further leads/improvements to follow at the moment, so no progress on the Atomic Bomberman front.

**Comment by jangatzke** [comment]
We had a LAN party with 10 players using 5 computers. The game got async using windows and the ipxwrapper version linked in this issue. We switched to Fedora + Wine + same wrapper version and the game was super stable and without lags.

**Comment by solemnwarning** [comment]
I'm not able to get Atomic Bomberman to run on any of my machines at the moment, but I've prototyped a rate limiting option into this build of IPXWrapper which may or may not help things:

[ipxwrapper-git.zip](https://github.com/user-attachments/files/20852308/ipxwrapper-git.zip)

Try experimenting with the packet and/or data rate limit options in ipxconfig and let me know how you get on (if you're still playing this game)!

**Comment by Ceotom** [comment]
If we set any packet rate limit game crashes with folowing error if we try to host or leave Join Network Game screen

<img width="414" height="273" alt="Image" src="https://github.com/user-attachments/assets/99842325-ab47-4e9c-a2f1-7775acce7696" />

**Comment by solemnwarning** [comment]
> If we set any packet rate limit game crashes with folowing error if we try to host or leave Join Network Game screen
> <img alt="Image" width="414" height="273" src="https://private-user-images.githubusercontent.com/36301121/468290331-99842325-ab47-4e9c-a2f1-7775acce7696.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI5NjU2MTksIm5iZiI6MTc1Mjk2NTMxOSwicGF0aCI6Ii8zNjMwMTEyMS80NjgyOTAzMzEtOTk4NDIzMjUtYWI0Ny00ZTljLWEyZjEtNzc3NWFjY2U3Njk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE5VDIyNDgzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlZDE4OWFiNDJhMTgzOGQ2ZjEwMmFhN2E1MzRjOWRkZWU5ZmY2MDZhZGRiYzIxMmUzNTYwNDA4ZDFkZmQyODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DKb19bdUI2mosQGBB_OqfuUiEJ1rmWnFhsG5ZO3L0Bg">

Whoops, thanks for pointing that out, should be fixed in this build: [ipxwrapper-git.zip](https://github.com/user-attachments/files/21331997/ipxwrapper-git.zip)

**Comment by Ceotom** [comment]
Coalesce packets: on
Rate limit packets: 100 packets/sec
Rate limit data: 10 KiB/sec

We figured out that these settings are the most optimal. Played for 3+ hours with 4 players with almost no issues, players movement is not smooth, like in local games, but game is playable now without desyncs, delays in other players actions and etc.

**Comment by insellium** [comment]
FINALLY

ATOMIC BOMBERMAN ONLINNE IS PLAYABLE NOW

**Comment by solemnwarning** [comment]
[IPXWrapper 0.7.2](https://github.com/solemnwarning/ipxwrapper/releases/tag/ipxwrapper-0.7.2) released.

**Comment by mayerwin** [comment]
Just adding here that to get acceptable performance, it is necessary to use IPXBox (to start a IPX server), as without it, the game would still lag too much (no smooth movement).

Not sure why using a server (running on the same machine as the game host) is less laggy than using IPXWrapper alone (connecting 2 computers with each other). 
This means there is probably an opportunity to look up what IPXBox is doing that IPXWrapper isn't in the way it handles packets, to improve IPXWrapper and make it unnecessary to use a IPX server.

I ended up publishing a guide to help future Bomberman users get started with multiplayer games as simply as possible:
https://github.com/mayerwin/BOMBERMAN-MULTIPLAYER/tree/main

**Comment by insellium** [comment]
> Not sure why using a server (running on the same machine as the game host) is less laggy than using IPXWrapper alone (connecting 2 computers with each other).

Did you set in IPXConfig a rate limits for packets and data?
for better ABM netplay you need to set 100 packets and 10 KiB.

**Comment by mayerwin** [comment]
> > Not sure why using a server (running on the same machine as the game host) is less laggy than using IPXWrapper alone (connecting 2 computers with each other).
> 
> Did you set in IPXConfig a rate limits for packets and data? for better ABM netplay you need to set 100 packets and 10 KiB.

Yes, I used the limits referenced in this thread. But it wasn't good enough without IPXBox.

**Comment by solemnwarning** [comment]
The packet coalescing is only implemented for DOSBox, so its probably that.

**Comment by mayerwin** [comment]
Ah indeed. Would it be hard to implement for non-DOSBox usages?

**Comment by solemnwarning** [comment]
No, its basically the same. I'll look into doing it when I find the time.

**Comment by solemnwarning** [comment]
Having thought about it a bit more, I don't think the coalesce option will actually be making a difference here - with the packet limit set at 100/sec it won't be sending enough to trigger it.

**Comment by Ceotom** [comment]
I didn't look at the logs, but it's actually makes noticeable difference.

**Comment by solemnwarning** [comment]
> I didn't look at the logs, but it's actually makes noticeable difference.

Well that's weird; packet coalescing isn't supposed to kick in until the application sends 512 packets within a 2.5 second window. Do you see the "High send rate to x detected, coalescing future packets" message in the log even with the packet limit set at 100/sec?

**Comment by Ceotom** [comment]
Nevermind, we're tested again and there's was no difference. Maybe we're tested this option with other combination of rate limit.

**Comment by Ceotom** [comment]
After testing with 10+ people on same DosBox server (in different sessions)  we're found that those rate limits makes game unstable with such amount of clients. Game always send multicast packets for no reason.

Yesterday we realized that DirectDraw tickrate (not FPS) affects amount of packets game sends. We're forced 60 ticks in DirectDraw wrapper and played few duels without any rate limit whatsoever and there's was no lag at all. Packet coalescing not triggered too. We will test with more people later. 

60 ticks also solves few other issues, so it's recommended in any case.
